### PR TITLE
🖊️ Patch Clinical Exception with Clinical Errors

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

- Removes migration errors which are covered by program / entity exceptions

**Type of Change**

- [x] Bug
- [ ] Refactor
- [ ] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [ ] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [ ] Manual testing
- [ ] Regression tests completed and passing (double check number of tests).
- [ ] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/clinical/clinical-service.ts [hunk 1]
```diff
@@ -289,13 +292,65 @@ export const getClinicalSearchResults = async (
...
+      );
+      // If no matching exceptions related to error, error is included in the result
+      const hasExceptions = matchedExceptions.length;
+      if (!hasExceptions) return errorRecord;
+
+      // Remove Error records where both field and value match the exception
+      const errors = errorRecord.errors.filter(error => {
+        const hasErrorValue = 'value' in error.info;
+
+        // Lectern client is not returning specific error values for every entry
```
Comment (demariadaniel) [inline src/clinical/clinical-service.ts:332]: ... but it should be, needs follow up ticket

Comment (demariadaniel) [reply src/clinical/clinical-service.ts:332]: See https://github.com/overture-stack/js-lectern-client/blob/master/src/schema-functions.ts#L629

Comment (joneubank) [reply src/clinical/clinical-service.ts:332]: Appreciate how far you've looked into this. Could you provide more explanation of what is wrong in the lectern client? Is this something we can fix in the client? More explanation of the problem would help understand the next action.

File: src/clinical/clinical-service.ts [hunk 2]
```diff
@@ -30,14 +31,16 @@ import {
   aliasEntityNames,
   allEntityNames,
 } from '../common-model/entities';
-import { Errors } from '../utils';
+import { Errors, notEmpty } from '../utils';
```
Comment (ciaranschutte) [inline src/clinical/clinical-service.ts:34]: `notEmpty` is unused

File: src/clinical/clinical-service.ts [hunk 3]
```diff
@@ -30,14 +31,16 @@ import {
...
 import { patchCoreCompletionWithOverride } from '../submission/submission-to-clinical/stat-calculator';
 import { migrationRepo } from '../submission/migration/migration-repo';
 import { MigrationManager } from '../submission/migration/migration-manager';
 import {
   DictionaryMigration,
   DonorMigrationError,
 } from '../submission/migration/migration-entities';
+import * as exceptionService from '../exception/exception-service';
+import { failure, Result, success, ValidationError } from '../exception/error-handling';
```
Comment (ciaranschutte) [inline src/clinical/clinical-service.ts:43]: these imports are unused

File: src/clinical/clinical-service.ts [hunk 4]
```diff
@@ -289,13 +292,65 @@ export const getClinicalSearchResults = async (
...
-  // 2. ...and now remove from the list all valid donors (fixed with submissions since the migration)
-  const validErrorRecords = await getValidRecordsPostSubmission(programId, migrationErrors);
+  // 2. Remove from the list all valid donors (fixed with submissions since the migration)
+  const validPostSubmissionErrors = await getValidRecordsPostSubmission(programId, migrationErrors);
+
+  // 3. Remove from the list all errors which match Program Exceptions
+  const programExceptions = await exceptionService.operations.getProgramException({ programId });
+  const entityExceptions = await exceptionService.operations.getEntityException({ programId });
+  const hasProgramExceptions = 'exception' in programExceptions;
+  const hasEntityExceptions = 'exception' in entityExceptions;
```
Comment (ciaranschutte) [inline src/clinical/clinical-service.ts:305]: `programExceptions` and `entityExceptions` are of types `Result<EntityException>` and `Result<ProgramException>` they'll have a `success` property which to me is more stable than checking for a key.
https://github.com/icgc-argo/argo-clinical/blob/develop/src/exception/error-handling.ts#L26

File: src/clinical/clinical-service.ts (hunk 3 above)
Comment (demariadaniel) [reply src/clinical/clinical-service.ts]: Now removed!

File: src/clinical/clinical-service.ts (hunk 4 above)
Comment (demariadaniel) [reply src/clinical/clinical-service.ts]: Yes but Jon pointed out that they were implemented as `success: boolean` vs `success: true` which was causing me major headaches, which was why I turned to `in`

I've updated this in https://github.com/icgc-argo/argo-clinical/pull/1091/commits/3c752f965ff3e194f7802322ad2495530f50f29d

The Errors function now just uses `checkForProgramAndEntityExceptions`

File: src/clinical/clinical-service.ts (hunk 2 above)
Comment (demariadaniel) [reply src/clinical/clinical-service.ts]: Using now; sorry PR needed to be revised

File: src/clinical/clinical-service.ts (hunk 1 above)
Comment (demariadaniel) [reply src/clinical/clinical-service.ts]: We found yesterday, Lectern Client wasn't truly causing an issue; but, there will be some cases where we may be unable to be 100% accurate labelling errors on the UI. When error info/value are not provided, we can leverage the `index` property, but will have to see how reliably that works.

**Comment by demariadaniel** [comment]
@joneubank @ciaranschutte this PR is now completely revised and working!

File: src/submission/exceptions/exceptions.ts [hunk 5]
```diff
@@ -35,7 +35,7 @@ import { DeepReadonly } from 'deep-freeze';
  * @param programId
  * @returns program and donor level exceptions for this programId
  */
-const queryForExceptions = async (programId: string) => {
+export const queryForExceptions = async (programId: string) => {
```
Comment (ciaranschutte) [inline src/submission/exceptions/exceptions.ts:38]: this doesn't need to be exported.

**Review by ciaranschutte — APPROVED** [review]

File: src/submission/exceptions/exceptions.ts (hunk 5 above)
Comment (demariadaniel) [reply src/submission/exceptions/exceptions.ts:38]: Right, forgot to remove from earlier dev work
Updated

**Review by ciaranschutte — APPROVED** [review]

**Pull request merged**